From 084bb0f4321fa5c5c0e2d840c3a0b9ce4419ffd8 Mon Sep 17 00:00:00 2001
From: Yong Zhi <yong.zhi@intel.com>
Date: Thu, 12 Apr 2018 10:28:45 -0500
Subject: [PATCH 1914/2302] media: intel-ipu4: ak7375: Add (un)subscribe_event
 core ops

Message for Open Source:
Fix v4l2-compliance fails on ak7375 lens controller subdev.

Message for Internal:
[Issue/Feature] v4l2-compliance -u /dev/v4l-subdev12

"test VIDIOC_(UN)SUBSCRIBE_EVENT/DQEVENT: FAIL"

[Root Cause/Changes]
Implementation following example of drivers/media/i2c/ov7740.c
(commit 65b0ae5e3e media: ov7740: Set subdev HAS_EVENT flag)

Change-Id: Ibcaec2a25a0100c4fecafa8eeaaf56d69c465159
Signed-off-by: Yong Zhi <yong.zhi@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/i2c/ak7375.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/media/i2c/ak7375.c b/drivers/media/i2c/ak7375.c
index a75eb4c..890983e 100644
--- a/drivers/media/i2c/ak7375.c
+++ b/drivers/media/i2c/ak7375.c
@@ -8,6 +8,7 @@
 #include <linux/pm_runtime.h>
 #include <media/v4l2-ctrls.h>
 #include <media/v4l2-device.h>
+#include <media/v4l2-event.h>
 
 #define AK7375_NAME		"ak7375"
 #define AK7375_MAX_FOCUS_POS	4095
@@ -108,7 +109,14 @@ static const struct v4l2_subdev_internal_ops ak7375_int_ops = {
 	.close = ak7375_close,
 };
 
-static const struct v4l2_subdev_ops ak7375_ops = { };
+static const struct v4l2_subdev_core_ops ak7375_subdev_core_ops = {
+	.subscribe_event = v4l2_ctrl_subdev_subscribe_event,
+	.unsubscribe_event = v4l2_event_subdev_unsubscribe,
+};
+
+static const struct v4l2_subdev_ops ak7375_ops = {
+	.core = &ak7375_subdev_core_ops,
+};
 
 static void ak7375_subdev_cleanup(struct ak7375_device *ak7375_dev)
 {
@@ -146,7 +154,8 @@ static int ak7375_probe(struct i2c_client *client,
 		return -ENOMEM;
 
 	v4l2_i2c_subdev_init(&ak7375_dev->sd, client, &ak7375_ops);
-	ak7375_dev->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE;
+	ak7375_dev->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE |
+				V4L2_SUBDEV_FL_HAS_EVENTS;
 	ak7375_dev->sd.internal_ops = &ak7375_int_ops;
 
 	rval = ak7375_init_controls(ak7375_dev);
-- 
2.7.4

